### PR TITLE
[DRAFT] Rewrite of the whole project with a toy model

### PR DESCRIPTION
**Pull request by noaonoszko**
I had ChatGPT generate a first draft of the code and now I'm working on
correcting it. I'll attach the prompt as it shows you my understanding of the
project and my plan for this rewrite. Please let me know if something looks
wrong.

"Give me a full code project that mocks all of this, and use dependency
injection to achieve this. I want to have these two classes:

- Miner: A node that trains
- Validator: A node that collects slices of the models after each communication
  round.

The algorithm will look like this:
- [x] 1. Models are initialized at each miner and at the validator. Either with
  identical weights or different, this should be governed by a variable.
- [x] 2. The miners get a page of the data (a subset), each of same size, to
  train on.
- [ ] 3. Each miner does one training step (i.e. one forward propagation, one
  backward prop and an update of the network.
- [ ] 4. Each miner sends a slice of the model to the validator. This slice is
  of size dataset_size/compression_factor. Which parameters that are sent is
  random and determined by the validator.
- [ ] 5. The validator averages the models. 
- [ ] 6. For each miner, calculate a score, or measure of how good their update
      was, defined as the loss of the validator model with the modification that
      the weights corresponding to the miner slice have been averaged with the
      miner's slice, minus the loss of the validator model.

      However, calculate this in four different ways, and use dependency injection to
      determine which way to use. Method 1 is exact calculation, i.e., do a forward
      prop for each of the two models and calculate the loss. Method 2 is by using a
      second order taylor expansion. Method 3 is first order taylor expansion. Method
      4 is first order taylor expansion, but the term is actually only the cosine
      similarity so we're just checking directionality. 

      Save this score so that we can plot it later.

- [ ] 7. Each miner gets a slice of the new averaged model. The slice is the
      same for all miners and the size of it is not necessarily the same as the
      size of the slice sent from the miner to the validator in step 4.

- [ ] 8. Continue from step 2 and redo this for num_communication_rounds times.

So as you can see, you need to abstract away the model and the dataset in some
nice way. Because I want to be able to do the same with LLMs later."

### CODE REVIEW

**Comment by noaonoszko** [comment]
This now exists here:
https://github.com/noaonoszko/boltzmann-experimentation
